kubernetes/resource_kubernetes_service_account.go

```diff
@@ -258,3 +269,80 @@ func resourceKubernetesServiceAccountExists(d *schema.ResourceData, meta interfa
 	}
 	return true, err
 }
+
+func resourceKubernetesServiceAccountImportState(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	conn := meta.(*kubernetes.Clientset)
+
+	namespace, name, err := idParts(d.Id())
+	if err != nil {
+		return nil, err
+	}
+
+	sa, err := conn.CoreV1().ServiceAccounts(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	defaultSecret, err := findDefaultServiceAccount(sa, conn)
+	if err != nil {
+		return nil, err
+	}
+
+	err = d.Set("default_secret_name", defaultSecret)
+	if err != nil {
+		return nil, err
+	}
+	d.SetId(buildId(sa.ObjectMeta))
+
+	return []*schema.ResourceData{d}, nil
+}
+
+func findDefaultServiceAccount(sa *api.ServiceAccount, conn *kubernetes.Clientset) (string, error) {
+	/*
+		The default service account token secret would have:
+		- been created either at the same moment as the service account or _just_ after (Kubernetes controllers appears to work off a queue)
+		- have a name starting with "[service account name]-token-"
+
+		See this for where the default token is created in Kubernetes
+		https://github.com/kubernetes/kubernetes/blob/release-1.13/pkg/controller/serviceaccount/tokens_controller.go#L384
+	*/
+	var serviceAccountTokens []string
+	for _, saSecret := range sa.Secrets {
+		if !strings.HasPrefix(saSecret.Name, fmt.Sprintf("%s-token-", sa.Name)) {
+			log.Printf("[DEBUG] Skipping %s as it doesn't have the right name", saSecret.Name)
+			continue
+		}
+
+		secret, err := conn.CoreV1().Secrets(sa.Namespace).Get(saSecret.Name, metav1.GetOptions{})
+		if err != nil {
+			return "", err
+		}
+
+		if secret.Type != api.SecretTypeServiceAccountToken {
+			log.Printf("[DEBUG] Skipping %s as it is of the wrong type", saSecret.Name)
+			continue
+		}
+
+		if secret.CreationTimestamp.Before(&sa.CreationTimestamp) {
+			log.Printf("[DEBUG] Skipping %s as it existed before the service account", saSecret.Name)
+			continue
+		}
+
+		if secret.CreationTimestamp.Sub(sa.CreationTimestamp.Time) > (1 * time.Second) {
+			log.Printf("[DEBUG] Skipping %s as it wasn't created at the same time as the service account", saSecret.Name)
+			continue
+		}
+
+		log.Printf("[DEBUG] Found %s as a candidate for the default service account token", saSecret.Name)
+		serviceAccountTokens = append(serviceAccountTokens, secret.Name)
+	}
+
+	if len(serviceAccountTokens) == 0 {
+		return "", fmt.Errorf("Unable to find any service accounts tokens which could have been the default one")
+	}
+
+	if len(serviceAccountTokens) > 1 {
```

Comment:
If the goal is to find exactly one, then finding the first ones means it's done.
You could just return it when you find it rather than running the list of secrets all the way to the end.

Reply:
The thought was to be as conservative as possible and just fail if the assumptions used to find the default service account aren't correct

Reply:
I'll refactor it so that is less conservative